### PR TITLE
Continuous Integration and Unit Testing

### PR DESCRIPTION
**Pull request by olgaboiar**
- Added CircleCI
- Added Test Coverage (Jacoco)
- Added Unit Tests

### CODE REVIEW

File: src/test/java/com/olgaboiar/guacomole/EchoClientTest.java [hunk 1]
```diff
@@ -15,11 +16,16 @@
     Socket socket;
 
     @BeforeEach
-    public void setup () throws Exception {
+    public void preTest () throws Exception {
```
Comment (naush) [inline src/test/java/com/olgaboiar/guacomole/EchoClientTest.java]: I'm curious why you choose the word preTest and postTest in favor of beforeEach/setUp and afterEach/tearDown?

File: src/main/java/com/olgaboiar/guacomole/EchoClient.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    int port;
+    Socket socket;
+    int connectionAttempts;
+    int maxConnectionAttempts;
+
+    public EchoClient(String host, int port) {
+        this.host = host;
+        this.port = port;
+        this.maxConnectionAttempts = 20;
+        this.connectionAttempts = 0;
```
Comment (naush) [inline src/main/java/com/olgaboiar/guacomole/EchoClient.java:18]: I would also make these two variables configurable.

File: src/main/java/com/olgaboiar/guacomole/EchoClient.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        socket = new Socket();
+        try {
+            socket.connect(new InetSocketAddress(host, port));
+        } catch (IOException e) {
+            if (this.connectionAttempts < this.maxConnectionAttempts) {
+                Thread.sleep(500);
+                this.connectionAttempts ++;
+                connect();
+            }
+        }
```
Comment (naush) [inline src/main/java/com/olgaboiar/guacomole/EchoClient.java:31]: Better! 🎉

File: src/main/java/com/olgaboiar/guacomole/EchoClient.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public byte[] send(byte[] data) throws IOException {
+        socket.getOutputStream().write(data);
+        byte[] readBuffer = new byte[data.length];
+        socket.getInputStream().read(readBuffer);
+        return readBuffer;
```
Comment (naush) [inline src/main/java/com/olgaboiar/guacomole/EchoClient.java:41]: Mildly violating single responsibility here. There are two major operations here, read and write.

**Comment by naush** [comment]
Could we put `build/reports/` or maybe `build` to .gitignore as well?

File: src/test/java/com/olgaboiar/guacomole/EchoClientTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    @AfterEach
+    public void postTest() throws Exception {
+        serverProcess.stop();
+    }
+
+    @Test
+    void testClientConnectsToServer () throws Exception {
+        socket = new Socket();
+        Thread.sleep(1000);
```
Comment (naush) [inline src/test/java/com/olgaboiar/guacomole/EchoClientTest.java]: Would be nice to not have to sleep.

**Comment by naush** [comment]
👍 A couple of minor issues. Let's start on http server next week!

File: src/test/java/com/olgaboiar/guacomole/EchoClientTest.java (hunk 1 above)
Comment (olgaboiar) [reply src/test/java/com/olgaboiar/guacomole/EchoClientTest.java]: fixed in https://github.com/olgaboiar/GuacamoleServer/pull/3/commits/33a874d6b5d5f52c7d47e4675a38e82c3c72f35a

File: src/main/java/com/olgaboiar/guacomole/EchoClient.java (hunk 4 above)
Comment (olgaboiar) [reply src/main/java/com/olgaboiar/guacomole/EchoClient.java:41]: refactored in https://github.com/olgaboiar/GuacamoleServer/pull/3/commits/33a874d6b5d5f52c7d47e4675a38e82c3c72f35a

File: src/test/java/com/olgaboiar/guacomole/EchoClientTest.java (hunk 5 above)
Comment (olgaboiar) [reply src/test/java/com/olgaboiar/guacomole/EchoClientTest.java]: solved in https://github.com/olgaboiar/GuacamoleServer/pull/3/commits/aae8861758b3375a3435f7e0d9be1c62448c5d7b

File: src/main/java/com/olgaboiar/guacomole/EchoClient.java (hunk 2 above)
Comment (olgaboiar) [reply src/main/java/com/olgaboiar/guacomole/EchoClient.java:18]: minor refactoring in https://github.com/olgaboiar/GuacamoleServer/pull/3/commits/aae8861758b3375a3435f7e0d9be1c62448c5d7b

**Pull request merged**